### PR TITLE
provider/openstack - refactor enable/disable ops to disable pool memb…

### PR DESCRIPTION
**Pull request by derekolk**
…ers instead of adding/removing them

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackLoadBalancerV2Provider.groovy [hunk 1]
```diff
@@ -32,7 +32,9 @@ import org.openstack4j.model.network.ext.ListenerV2
 import org.openstack4j.model.network.ext.LoadBalancerV2
 import org.openstack4j.model.network.ext.LoadBalancerV2StatusTree
 import org.openstack4j.model.network.ext.MemberV2
+import org.openstack4j.model.network.ext.MemberV2Update
 import org.openstack4j.model.network.ext.Protocol
+import org.openstack4j.model.network.ext.builder.MemberV2UpdateBuilder
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackLoadBalancerV2Provider.groovy]: I don't think you are using this import.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy [hunk 2]
```diff
@@ -54,16 +57,16 @@ abstract class AbstractEnableDisableOpenstackAtomicOperation implements AtomicOp
   @Override
   Void operate(List priorOutputs) {
     String verb = disable ? 'disable' : 'enable'
-    String presentParticipling = disable ? 'Disabling' : 'Enabling'
+    String gerund = disable ? 'Disabling' : 'Enabling'
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy]: What does `gerund` mean?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy [hunk 3]
```diff
@@ -75,6 +78,51 @@ abstract class AbstractEnableDisableOpenstackAtomicOperation implements AtomicOp
     }
   }
 
+  void enableDisableLoadBalancerMembers(List<String> instanceIds, List<String> loadBalancerIds) {
+    String gerund = disable ? 'Disabling' : 'Enabling'
+    task.updateStatus phaseName, "Deregistering instances from load balancers..."
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy]: Are you always deregistering instances? Doesn't this also register?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy [hunk 4]
```diff
@@ -16,64 +16,120 @@
...
+import com.netflix.spinnaker.clouddriver.openstack.provider.OpenstackInfrastructureProvider
 import groovy.transform.Canonical
+import org.openstack4j.model.compute.Address
 import org.openstack4j.model.compute.Server
 
-import static org.openstack4j.model.compute.Server.Status
-
 @Canonical
-class OpenstackInstance implements Instance, Serializable {
+class OpenstackInstance {
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy]: Why no longer implement `Instance`?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy [hunk 5]
```diff
@@ -16,64 +16,120 @@
...
-        HealthState.Starting
-        break
-      case Status.UNKNOWN:
-        HealthState.Unknown
-        break
-      default:
-        HealthState.Down
-    }
-  }
+  String ipv6
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy]: You added `ipv6` but I don't see a field for the ipv4 address.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy [hunk 6]
```diff
@@ -16,64 +16,120 @@
...
+  List<OpenstackLoadBalancerHealth> loadBalancerHealths = []
+  OpenstackInstanceHealth instanceHealth
 
   static OpenstackInstance from(Server server, String account, String region) {
+    Address fixedIpAddress = server?.addresses?.getAddresses('private')?.find { it.type == 'fixed' && it.version == 6 }
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy]: What is this used for?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy [hunk 7]
```diff
@@ -129,6 +140,45 @@ class OpenstackLoadBalancerCachingAgent extends AbstractOpenstackCachingAgent im
         //create load balancer. Server group relationships are not cached here as they are cached in the server group caching agent.
         OpenstackLoadBalancer openstackLoadBalancer = OpenstackLoadBalancer.from(loadBalancer, resultlisteners, pool, healthMonitor, accountName, region)
 
+        // Populate load balancer healths and find instance ids which are members of the current lb via membership
+        List<OpenstackLoadBalancerHealth> healths = []
+        List<String> instanceKeys = []
+
+        LoadBalancerV2StatusTree loadBalancerV2StatusTree = clientProvider.getLoadBalancerStatusTree(region, loadBalancer.id)
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: Can we prefetch all of the status trees via CompletableFuture.supplyAsync { get loadbalancers }.thenConsume { get status tree of each can cache it in a map }

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+import static org.openstack4j.model.compute.Server.Status
+
+class OpenstackInstanceHealth {
+  Status status
+
+  HealthState toHealthState() {
+    switch (status) {
+      case Status.ACTIVE:
+        HealthState.Unknown
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy]: Should this be `Active` not `Unknown`?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy [hunk 9]
```diff
@@ -129,6 +140,45 @@ class OpenstackLoadBalancerCachingAgent extends AbstractOpenstackCachingAgent im
...
+
+        Map<String, String> memberStatusMap = loadBalancerV2StatusTree?.loadBalancerV2Status?.listenerStatuses?.collectEntries { ListenerV2Status listenerStatus ->
```
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: I think this needs to be encapsulated somewhere, so we can call it to get the overall servergroup healthStatus.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        HealthState.Starting
+        break
+      case Status.ERROR:
+        HealthState.Failed
+        break
+      case Status.UNKNOWN:
+        HealthState.Unknown
+        break
+      default:
+        HealthState.Down
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy]: Seems odd to me that `default` maps to `Down` not `Unknown`. Does Openstack not send a `Down` state?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+  List<LBHealthSummary> lbHealthSummaries
+  PlatformStatus status
+
+  enum PlatformStatus {
+    ONLINE,
+    DISABLED
+
+    HealthState toHealthState() {
+      switch(this) {
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy]: Could be simplified to `return this == ONLINE ? Up : Down`

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        case ONLINE:
+          return HealthState.Up
+        default:
+          return HealthState.Down
+      }
+    }
+
+    LBHealthSummary.ServiceStatus toServiceStatus() {
+      switch (this) {
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy]: Could be simplified to `return this == ONLINE ? InService :  OutOfService`

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy [hunk 13]
```diff
@@ -129,6 +140,45 @@ class OpenstackLoadBalancerCachingAgent extends AbstractOpenstackCachingAgent im
...
+          listenerStatus.lbPoolV2Statuses?.collectEntries { LbPoolV2Status poolStatus ->
+            poolStatus.memberStatuses?.collectEntries { MemberV2Status memberStatus ->
+              [memberStatus.address, memberStatus.operatingStatus]
+            }
+          }
+        }
+
+        // Read instances from cache and create a map indexed by ipv6 address to compare to load balancer member status
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: Why only by ipv6? What if they don't have a v6 addr?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackLoadBalancerV2Provider.groovy (hunk 1 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackLoadBalancerV2Provider.groovy]: Done

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy (hunk 2 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy]: https://en.wikipedia.org/wiki/Gerund

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy (hunk 3 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/AbstractEnableDisableOpenstackAtomicOperation.groovy]: Done

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy (hunk 4 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy]: The View class does that.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy (hunk 5 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy]: Members always use ipv6.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy (hunk 6 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstance.groovy]: Helping to correlate instances with members (which only have ips, not instance ids)

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy (hunk 8 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy]: No - it is needed in order for operations to pass the orca checks for enabling/disabling. It's odd, but this is how Google is doing it too.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy (hunk 10 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackInstanceHealth.groovy]: probably

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy (hunk 11 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy]: yes

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy (hunk 12 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/model/OpenstackLoadBalancerHealth.groovy]: yes

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy (hunk 13 above)
Comment (drmaas) [reply clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/provider/agent/OpenstackLoadBalancerCachingAgent.groovy]: again, members only store the v6 address

**Review by drmaas — APPROVED** [review]

**Pull request closed without merging**